R/do_findChromPeaks-functions.R

```diff
@@ -2220,9 +2220,220 @@ do_findPeaks_MSW <- function(mz, int, snthresh = 3,
+                                           snthresh = 10,
+                                           prefilter = c(3, 100),
+                                           mzCenterFun = "wMean",
+                                           integrate = 1,
+                                           mzdiff = -0.001,
+                                           fitgauss = FALSE,
+                                           noise = 0,
+                                           verboseColumns = FALSE,
+                                           roiList = list(),
+                                           firstBaselineCheck = TRUE,
+                                           roiScales = NULL,
+                                           sleep = 0,
+                                           extendLengthMSW = FALSE,
+                                           ppmMerging = 10,
+                                           binWidthIM = 0.01
+                                          ){
+    ## Extract frame information
+    pdata <- peaksData(spec, columns = c("mz", "intensity"))
+    rt <- rtime(spec)
+    im <- spec$inv_ion_mobility
+    
+    
+    ## Merging frames into scans and Summarize across IM dimension
+    message("Collapsing data over IM dimension... ", appendLF = F)
+    
+    scans_summarized <-
+        combineSpectra(
+            spec,
+            f = as.factor(spec$frameId),
+            intensityFun = base::sum,
+            weighted = TRUE,
+            ppm = ppmMerging
+        )
+    message("OK")
+    
+    ## Peak-picking on summarized data
+    mzs <- mz(scans_summarized)
+    valsPerSpect <- lengths(mzs, FALSE)
+    mz <- unlist(mzs, use.names = FALSE)
+    int <- unlist(intensity(scans_summarized), use.names = FALSE)
+    scantime <- sort(unique(rt))
+    peaks <- .centWave_orig(mz = mz, int = int, scantime = scantime,
+                   valsPerSpect = valsPerSpect, ppm = ppm, peakwidth = peakwidth,
+                   snthresh = snthresh, prefilter = prefilter,
+                   mzCenterFun = mzCenterFun, integrate = integrate,
+                   mzdiff = mzdiff, fitgauss = fitgauss, noise = noise,
+                   verboseColumns = verboseColumns, roiList = roiList,
+                   firstBaselineCheck = firstBaselineCheck,
+                   roiScales = roiScales, sleep = sleep,
+                   extendLengthMSW = extendLengthMSW)
+    
+    ## Resolving peaks across IM dimension
+    message("Resolving peaks over ion-mobility dimension... ", appendLF = F)
+    resolved_peaks <- vector("list", nrow(peaks))
+    for (i in seq_len(nrow(peaks))) {
+        current_peak <- peaks[i,]
+        mobilogram <- .extract_mobilogram(pdata, current_peak, rt, im, binWidthIM)
+        if (length(mobilogram) == 0) {
+            warning(i, " mobilogram is empty")
+            next
+        }
+        bounds <- .split_mobilogram(mobilogram)
+        new_peaks <- data.frame(
```

Comment:
I think using `data.frame` here might have an impact on performance. I think `matrix` might be faster, but that's something we can check later.